### ISSUE TITLE
[Validation] v0.9 reference + v0.10 Triton: lm_eval, HF ecosystem, and SFT/DPO/GRPO evidence

### ISSUE THREAD

**Issue by 123123213weqw** [rwkv-rs/hf-adapter#144]
## 这条 Issue 记录什么

这是一份当前可复现状态报告，覆盖两条彼此独立的实现：

1. **v0.9.0 reference**：`main` 上发布的纯 PyTorch、可阅读、兼容性优先实现；
2. **v0.10 optional Triton（开发中）**：独立性能分支上的可选推理后端。

这里把“HF 能运行”“`lm_eval` 指标稳定”“训练链路能反向更新”和“优化 kernel 已通过严格数值晋级”分开记录。前三项已经有实证；最后一项目前仍未通过，因此不能把开发中的 Triton 后端描述为已经完成数值晋级。

---

## 结论摘要

| 验证项 | v0.9 reference | v0.10 forced Triton |
|---|---:|---:|
| 标准 HF AutoModel / cache / generation / save-reload | PASS | PASS |
| 正式 `lm_eval==0.4.9.1` | V100：48/48 | V100 / RTX 4080 / RTX 4090：均 48/48 |
| batch 1/8 指标稳定性 | PASS | 三卡均 PASS |
| LoRA SFT | V100 100 steps PASS | 三卡 1-step smoke PASS |
| LoRA DPO | V100 100 steps PASS | 三卡 1-step smoke PASS |
| LoRA GRPO | V100 100 steps PASS | 三卡 1-step smoke PASS |
| 有限 loss、参数更新、adapter 重载 | PASS | PASS |
| checkpoint resume | 100 → 101 PASS | 训练 fallback smoke PASS |
| W&B | offline smoke PASS；默认关闭 | 非必需；本地 JSON 为准 |
| Triton 加速训练 backward | 不适用 | **未实现；训练自动回退 PyTorch autograd** |
| 严格 FP16 full-model logits `max_abs <= 0.15` | reference/official oracle PASS | **三卡均有少数 case 未通过** |
| 性能后端发布状态 | 不适用 | **尚未晋级、尚未合入 main** |

因此当前准确表述是：

> v0.9.0 已通过正式 HF 参考实现的评测与训练门禁。可选 Triton 后端在 V100、RTX 4080、RTX 4090 上均能完成完整 `lm_eval` 和 HF 工作流；安装性能后端后训练仍可工作，但训练使用参考 autograd fallback，而不是 Triton backward。Triton 的少量 FP16 full-model logits 绝对误差仍超过严格阈值，所以性能分支继续保留 `passed: false`，不进入 `main`。

---

## 1. v0.9.0 reference：正式 `lm_eval`

固定配置：

- `lm_eval==0.4.9.1`
- 模型：0.1B、0.4B、1.5B
- batch size：1、8
- `max_length=2048`
- FP16、0-shot
- 不使用 `--limit`
- 任务：
  - `wikitext`
  - `lambada_openai`
  - `piqa`
  - `hellaswag`
  - `winogrande`
  - `arc_easy`
  - `arc_challenge`
  - `openbookqa`

总计：`3 models × 2 batch sizes × 8 tasks = 48 units`。

结果：

- 48/48 正常退出；
- 无 NaN / Inf；
- batch 1/8 普通指标绝对差 `<= 0.001`；
- Wikitext perplexity 相对差 `<= 0.1%`；
- validator：`status: passed`，`failures: []`。

最终 bundle 透明保留了复合来源：原正式轮保留 42 个未受影响单元；batch regrouping 修复后，只重跑受影响的 1.5B PIQA、OpenBookQA、WinoGrande batch 1/8 共六个单元。每个单元都记录代码 SHA、模型 revision、task config、dataset fingerprint、命令和退出状态。

证据：

- [V100 formal lm_eval composite](https://github.com/rwkv-rs/hf-adapter/tree/2eac22e5f05b3350da4e23a28322268a2a4ce15e/results/lm_eval/v100-composite-cfeb5ae)
- [48-unit validation.json](https://github.com/rwkv-rs/hf-adapter/blob/2eac22e5f05b3350da4e23a28322268a2a4ce15e/results/lm_eval/v100-composite-cfeb5ae/validation.json)
- [评测命令与说明](https://github.com/rwkv-rs/hf-adapter/blob/2eac22e5f05b3350da4e23a28322268a2a4ce15e/docs/EVALUATION.md)
- [正式矩阵 validator](https://github.com/rwkv-rs/hf-adapter/blob/2eac22e5f05b3350da4e23a28322268a2a4ce15e/evaluation/validate_lm_eval_matrix.py)

### reference 数学正确性

`lm_eval` 之外，参考实现还直接对照官方 RWKV token-wise equations，而不是把 FLA 当 correctness oracle：

| 硬件 | 模型 | dtype | 结果 |
|---|---|---|---:|
| Tesla V100 32GB | 0.1B / 0.4B / 1.5B | FP32 / FP16 | 6/6 PASS |
| RTX 4080 16GB | 0.1B / 0.4B / 1.5B | FP32 / FP16 / BF16 | 9/9 PASS |

覆盖 B=1/4、T=1/17/128、最终 state、teacher-forced cached decode、64-token greedy 和 loss。

证据：[official checkpoint oracle summary](https://github.com/rwkv-rs/hf-adapter/blob/2eac22e5f05b3350da4e23a28322268a2a4ce15e/results/official/README.md)

---

## 2. v0.9.0 reference：训练与 RL 验证

三套示例都是直接调用 TRL/PEFT，不是仓库内部的私有 Trainer 包装：

| 方法 | 数据集 | 固定数据 revision | 正式步数 | 结果 |
|---|---|---|---:|---:|
| LoRA SFT | `HuggingFaceH4/ultrachat_200k` | `8049631c405ae6576f93f445c6b8166f76f5505a` | 100 | PASS |
| LoRA DPO | `HuggingFaceH4/ultrafeedback_binarized` | `3949bf5f8c17c394422ccfab0c31ea9c20bdeb85` | 100 | PASS |
| LoRA GRPO | `openai/gsm8k` | `740312add88f781978c0658806c59bc2815b9866` | 100 | PASS |

共同配置：

- 0.1B 模型、seed 42、长度 512；
- LoRA `r=8 / alpha=16 / dropout=0.05`；
- target：`r_proj,k_proj,v_proj,o_proj,key,value`；
- `use_cache=False`，支持 gradient checkpointing；
- 保存 resolved config、环境、数据 fingerprint、确定性 sample indices、JSONL metrics、退出状态和 checkpoint SHA256。

三条正式训练均满足：

- loss 有限；
- 梯度非零；
- 144 个 LoRA 参数项发生变化；
- adapter save/reload logits `max_abs=0.0`；
- SFT/DPO/GRPO 的最终 validator 均为 PASS；
- SFT 从 checkpoint 100 恢复到 101，global step 正常推进；
- W&B offline smoke 通过（run ID `wy328yot`），但本地文件始终是事实来源，W&B 默认关闭。

环境：Tesla V100 32GB，PyTorch 2.5.1+cu124，Transformers 4.56.2，TRL 0.20.0，PEFT 0.19.1。由于该 Transformers 版本对 optimizer state 的安全恢复门禁，resume 使用 PyTorch 2.6.0+cu124。

证据：

- [V100 canonical fine-tuning summary](https://github.com/rwkv-rs/hf-adapter/blob/2eac22e5f05b3350da4e23a28322268a2a4ce15e/results/finetune/v100-reference-20260826/README.md)
- [SFT/DPO/GRPO validation.json](https://github.com/rwkv-rs/hf-adapter/blob/2eac22e5f05b3350da4e23a28322268a2a4ce15e/results/finetune/v100-reference-20260826/validation.json)
- [可直接运行的训练说明](https://github.com/rwkv-rs/hf-adapter/blob/2eac22e5f05b3350da4e23a28322268a2a4ce15e/docs/FINETUNING.md)
- [训练生态 CI smoke](https://github.com/rwkv-rs/hf-adapter/blob/2eac22e5f05b3350da4e23a28322268a2a4ce15e/tests/test_training_ecosystem.py)

---

## 3. v0.10 optional Triton：三卡强制路由验证

性能线固定：

```bash
RWKV7_BACKEND=optimized
RWKV7_KERNEL_IMPL=triton
```

共同代码：`cfeb5aeca860ce444ebb3515a20cc22f7e2b090b`  
实际 route：`native-triton-rank1-scan-v1`  
kernel wheel SHA256：`544b08225351d9658a1941c260a70710d47bb677d2976d730cec9e4f34771560`

注意：`graph` 只是 PyTorch reference recurrence + CUDA Graph 的兼容基线；只有上述 `triton` route 才计入这里的性能后端结果。

### 三卡矩阵

| Gate | V100 | RTX 4080 | RTX 4090 |
|---|---:|---:|---:|
| 实际 Triton route 核对 | PASS | PASS | PASS |
| 正式 `lm_eval` | 48/48 | 48/48 | 48/48 |
| batch 1/8 metric validator | PASS | PASS | PASS |
| HF AutoModel / generation / beam / save-reload / cache | PASS | PASS | PASS |
| 训练 reference fallback | PASS | PASS | PASS |
| LoRA SFT/DPO/GRPO smoke | PASS | PASS | PASS |
| adapter reload | `max_abs=0.0` | `max_abs=0.0` | `max_abs=0.0` |
| speed job | PASS | PASS | PASS |
| FLA speed | fused recurrent PASS | PASS | PASS |
| FLA operator/state/64-token greedy | fused recurrent PASS | PASS | PASS |
| 严格 FP16 model-logit gate | FAIL | FAIL | FAIL |
| 最终 numerical promotion | NOT PROMOTED | NOT PROMOTED | NOT PROMOTED |

每张卡的 Triton `lm_eval` 都是完整 48 单元，并通过 batch 1/8 合并校验。三张卡的 SFT/DPO/GRPO smoke 都完成 1 step、72 个参数项发生变化、adapter 重载误差为 0。

证据：

- [三卡汇总](https://github.com/123123213weqw/hf-adapter/blob/0c5ea30ac6868974ba9836c4a065fa8b2847af68/results/native-backend/TRITON_CFEB5AE_SUMMARY.md)
- [V100 compact bundle](https://github.com/123123213weqw/hf-adapter/tree/0c5ea30ac6868974ba9836c4a065fa8b2847af68/results/native-backend/v100-cfeb5ae-triton)
- [RTX 4080 compact bundle](https://github.com/123123213weqw/hf-adapter/tree/0c5ea30ac6868974ba9836c4a065fa8b2847af68/results/native-backend/4080-cfeb5ae-triton)
- [RTX 4090 compact bundle](https://github.com/123123213weqw/hf-adapter/tree/0c5ea30ac6868974ba9836c4a065fa8b2847af68/results/native-backend/4090-cfeb5ae-triton)

### 为什么训练能工作，但不能说 Triton 加速了训练

当前 Triton rank-1 scan 是推理实现。需要 autograd/backward、gradient checkpointing 或训练 loss 时，dispatcher 会回退到可微的纯 PyTorch reference op。因此：

```text
安装可选性能后端
  ├─ 支持的 FP16 inference → native-triton-rank1-scan-v1
  └─ training / backward   → reference PyTorch autograd fallback
```

这证明“安装性能包不会破坏 Trainer/PEFT/TRL”，但不等于“Triton 已经加速训练”。后者需要独立实现和验证 backward kernel，当前没有这样的结论。

### 尚未通过的数值门禁

三卡的最终 JSON 都保留 `passed: false`。原因不是 `lm_eval`、cache、greedy 或训练失败，而是少数 optimized-vs-reference FP16 full-model logits 的 `max_abs` 超过预设 `0.15`：

- 所有值有限；
- logits/state cosine 很高；
- cache 与 teacher-forced decode 通过；
- 64-token greedy 完全一致；
- batch regrouping 通过；
- 48/48 `lm_eval` 及 batch 稳定性通过；
- 但严格 max-abs gate 未通过。

4080/4090 的 clean-vs-FLA 也是仅 full-model logits gate 未过，operator、state 和 greedy 通过。V100 的 pinned FLA chunk/backward 在 sm70 上无法由测试的 Triton toolchain 编译，因此 V100 只保留 fused-recurrent inference parity/speed；这属于外部 FLA/toolchain 限制，报告中没有隐藏。

在 kernel 修正或门限经过有依据的校准之前，性能分支不合入 `main`，也不发布为稳定默认后端。

---

### 已证明

- 自包含 HF 模型能用标准 `AutoModelForCausalLM` 加载；
- cache、padding、generation、beam、save/reload、causal loss 和 gradient checkpointing 可工作；
- v0.9 reference 能完成正式评测和真实 100-step SFT/DPO/GRPO；
- 可选 Triton 推理 route 在 V100/4080/4090 上均完成正式 48-unit `lm_eval`；
- 安装性能后端后，训练框架能安全走 reference backward fallback；
- PEFT adapter 能保存并无误差重载。

### 尚未证明

- Triton backward 或“高性能训练”；
- 所有 GPU、所有 dtype、所有模型尺寸和所有分布式拓扑；
- vLLM/SGLang/verl 原生 rollout backend；
- 当前 Triton FP16 数值门禁已完成晋级。

这几个边界不能用“HF 接口兼容”自动推导，需要各自单独实现和验收。

---

## 5. 发布状态

- [GitHub v0.9.0 release](https://github.com/rwkv-rs/hf-adapter/releases/tag/v0.9.0)
- [PyPI `rwkv7-hf==0.9.0`](https://pypi.org/project/rwkv7-hf/0.9.0/)
- [六个公开 HF 模型及固定 revision](https://github.com/rwkv-rs/hf-adapter/blob/2eac22e5f05b3350da4e23a28322268a2a4ce15e/docs/PUBLISHED_MODELS.md)
- [PyPI wheel 回装与 V100 Hub smoke](https://github.com/rwkv-rs/hf-adapter/tree/2eac22e5f05b3350da4e23a28322268a2a4ce15e/results/release/wheel-0.9.0)

这条 Issue 后续只追加重新运行得到的、带代码 SHA/环境/命令/JSON 证据的结果；不会把接口层面的“理论可接入”写成已经跑过。

**Comment by 123123213weqw** [comment]
可选优化后端的功能范围、三卡 `lm_eval`/训练生态状态以及 Reference/FLA 全 shape 速度矩阵已单独整理到 #145。这里继续只保留 v0.9.0 clean reference 的发布验证。
